### PR TITLE
Bisaacson ​[Project 02] First Last - Triangle Drawing

### PR DESCRIPTION
**Pull request by Brooken2**
screenshot 51

### CODE REVIEW

File: pipeline.cpp [hunk 1]
```diff
@@ -71,6 +76,39 @@ void DrawPoint(Buffer2D<PIXEL> & target, Vertex* v, Attributes* attrs, Attribute
...
+}
+
+/*******************************************************
+ * crossProduct
+ *   By taking the two vertex point we can times them together and
+ *   Perform the cross product of the two and return the float
+ * ******************************************************/
+float crossProduct(Vertex one, Vertex two)
+{
+    float crossPro = 0.0;
```
Comment (scott56gps) [inline pipeline.cpp]: Good job in declaring a float before assigning it.  I see this helps in ensuring what return value will result from the operation.

I think the Vertex class uses Doubles for its attributes.  If this is the case, there is the possibility of data loss going from a double to a float.

File: pipeline.cpp [hunk 2]
```diff
@@ -80,7 +118,45 @@ void DrawLine(Buffer2D<PIXEL> & target, Vertex* const triangle, Attributes* cons
  ************************************************************/
 void DrawTriangle(Buffer2D<PIXEL> & target, Vertex* const triangle, Attributes* const attrs, Attributes* const uniforms, FragmentShader* const frag)
 {
-    // Your code goes here
+
+    //By finding the max and min of both are x and y axis we can find the outer box of the triangle. 
+    int maxX = fmax((int)triangle[0].x, fmax((int)triangle[1].x, (int)triangle[2].x));
```
Comment (scott56gps) [inline pipeline.cpp]: There is also the MAX3 function defined in definitions.  You got fmax to work fine here.

File: pipeline.cpp [hunk 3]
```diff
@@ -80,7 +118,45 @@ void DrawLine(Buffer2D<PIXEL> & target, Vertex* const triangle, Attributes* cons
...
+            float s = (float)crossProduct(q, vs2) / crossProduct(vs1, vs2);
+            float t = (float)crossProduct(vs1, q) / crossProduct(vs1, vs2);
+
+            //If we are in the traingle these conditions will be true
+            if ( (s >= 0) && (t >= 0) && (s + t <= 1))
+            {
+                //set the point that we are at to send to the draw point
+                Vertex myPoint[1];
+                myPoint[0] = { x , y };
+                DrawPoint(target, myPoint, attrs, uniforms, frag);
```
Comment (scott56gps) [inline pipeline.cpp]: I simply set the target[y][x] value here.  But what you did in calling DrawPoint is fine.

Comment (scott56gps) [reply pipeline.cpp]: In fact, I forgot that there is a DrawPoint function to use, so good job in using your resources!

File: pipeline.cpp [hunk 4]
```diff
@@ -80,7 +118,46 @@ void DrawLine(Buffer2D<PIXEL> & target, Vertex* const triangle, Attributes* cons
...
+        for (int y = minY; y <= maxY; y++)
+        {
+            //Our original vertex that we take the  x and y positing that we are at in the box
+            Vertex q;
+            q.x = (x - (int)triangle[0].x);
+            q.y = (y - (int)triangle[0].y);
+
+            //We find the cross product of each part and then divid them to find were we are if we are in the triangle
+            float s = (float)crossProduct(q, vs2) / crossProduct(vs1, vs2);
+            float t = (float)crossProduct(vs1, q) / crossProduct(vs1, vs2);
```
Comment (fizhy37) [inline pipeline.cpp]: Nice job making the cross product into its own function. It makes everything look a lot nicer

**Pull request merged**